README.md
# Regdown

![Build Status](https://github.com/cfpb/regdown/workflows/ci/badge.svg)[![Coverage Status](https://coveralls.io/repos/cfpb/regdown/badge.svg)](https://coveralls.io/r/cfpb/regdown)

Regdown is a [Python-Markdown](https://python-markdown.github.io) extension for interactive regulation text.

- [Dependencies](#dependencies)
- [Installation](#installation)
- [Documentation](#documentation)
- [Getting help](#getting-help)
- [Getting involved](#getting-involved)
- [Licensing](#licensing)
- [Credits and references](#credits-and-references)

## Dependencies

- Python 3.6, 3.8
- [Python-Markdown](https://python-markdown.github.io) 3.2

## Installation

First, install Regdown:

```shell
pip install regdown
```

Then you can either:


- Specify Regdown as an extension in calls to `markdown`:

    ```python
    import markdown
    from regdown import RegulationsExtension

    markdown.markdown(text, extensions=[RegulationsExtension()],)
    ```

- Use convenience `regdown` function to render Markdown with the `RegulationsExtension`:

    ```python
    from regdown import regdown

    regdown(text)
    ```

## Documentation


Regdown adds three major features to Markdown to support making federal regulations easier to navigate and read.

### Labeled Paragraphs

`{label} Paragraph text`

Each paragraph can have a defined label, using `{label}` syntax at the start of the paragraph. This is translated into an `id` attribute on the resulting HTML paragraph element. If no label is given, the contents of the paragraph are hashed to generate a unique `id` for that paragraph. This makes any paragraph in the text directly linkable.

### Pseudo Forms

- `Form field: __`
- `__Form Field`
- `inline__fields__`

Example print forms, where the `\_\_` indicate a space for hand-written input. Can be any number of underscores between 2 and 50.

### Section symbols

`§ 1024.5(d)`
`§1024.5(d)`

Section symbols will always have a non-breaking space (&nbsp;) inserted between them and whatever follows to avoid hanging a symbol at the end of a line.

### Block references

`see(label)`

Insert the contents of labeled paragraphs in other Regdown documents inline into the current document. 

References can be placed before or after paragraphs. These references are to labeled paragraphs in other Markdown documents. When a `contents_resolver` callback and `url_resolver` callback are provided, the text of those other paragraphs can be looked up and inserted inline into the document making the reference. If `render_block_reference` callback is provided, custom rendering of the referenced text to HTML can be performed.

Callbacks:

- `contents_resolver(label)`: resolve the paragraph label and return the Markdown contents of that paragraph if the paragraph exists.
- `url_resolver(label)`: resolve the paragraph label and return a URL to that paragraph if the paragraph exists.
- `render_block_reference(contents, url=None)`: render the contents of a block reference to HTML. The url to the reference may be give as a keyword argument if `url_resolver` is provided.

```python
from regdown import regdown

def my_contents_resolver(label):
    # Lookup the document that contains the given label …
    return corresponding_markdown_text

def my_block_renderer(block_markdown_contents, url=None):
    # Render the block to HTML
    return block_html

regdown(
    text,
    contents_resolver=my_contents_resolver,
    render_block_reference=my_block_renderer
)
```

## Getting help

Please add issues to the [issue tracker](https://github.com/cfpb/regdown/issues).

## Getting involved

General instructions on _how_ to contribute can be found in [CONTRIBUTING](CONTRIBUTING.md).

## Licensing
1. [TERMS](TERMS.md)
2. [LICENSE](LICENSE)
3. [CFPB Source Code Policy](https://github.com/cfpb/source-code-policy/)

## Credits and references

regdown was forked from [Wagtail-Flags](https://github.com/cfpb/wagtail-flags), which was itself forked from [consumerfinance.gov](https://github.com/cfpb/consumerfinance.gov).


CHANGELOG.md
## 1.0.5

- Replace the use of deprecated markdown.util.etree
- Changed Python 3.7 to 3.10
- Pin markdown under 3.5

## 1.0.4

- Include attr_list, allowing regdown users to set html attributes

## 1.0.3

- Replace Travis with GitHub Actions and Replace Travis Badge with Github Action Badge
- Remove unnecessary dependency on sha3 package

## 1.0.2

- Update for Markdown 3.2, drop 3.0 support #2
- Changed em_strong to em_strong2
- Dropped support for Python 2.7
- Changed Python 3.7 to 3.8

## 1.0.1

- Fix the HTML em dash entity [#1](https://github.com/cfpb/regdown/pull/1)

## 1.0.0

- Initial release! 


regdown/tests.py
# -*- coding: utf-8 -*-
import unittest

import markdown

from regdown import (
    DEFAULT_RENDER_BLOCK_REFERENCE,
    extract_labeled_paragraph,
    regdown,
)


class RegulationsExtensionTestCase(unittest.TestCase):
    def test_label(self):
        text = "{my-label} This is a paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-1" '
            'data-label="my-label" id="my-label">'
            "This is a paragraph with a label.</p>",
        )

    def test_nolabel(self):
        text = "This is a paragraph without a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block" '
            'data-label="" '
            'id="e2cb7f25f263e65fc6737e03e0ecb90382398da3966b6da734b451be">'
            "This is a paragraph without a label.</p>",
        )

    def test_linebreak_label(self):
        text = "{my-label}\nThis is a paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-1" '
            'data-label="my-label" id="my-label">'
            "This is a paragraph with a label.</p>",
        )

    def test_multiple_linebreaks_label(self):
        text = "{my-label}\n\nThis is a paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<div class="regdown-block level-1" data-label="my-label" '
            'id="my-label"></div>\n'
            '<p class="regdown-block" data-label="" '
            'id="725445113243d57f132b6408fa8583122d2641e591a9001f04fcde08">'
            "This is a paragraph with a label.</p>",
        )

    def test_inline_interp_label(self):
        text = "{1-a-Interp-1}\nThis is a paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-0" '
            'data-label="1-a-Interp-1" id="1-a-Interp-1">'
            "This is a paragraph with a label.</p>",
        )

    def test_deeper_inline_interp_label(self):
        text = "{1-a-Interp-1-i}\nThis is a paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-1" '
            'data-label="1-a-Interp-1-i" id="1-a-Interp-1-i">'
            "This is a paragraph with a label.</p>",
        )

    def test_even_deeper_inline_interp_label(self):
        text = "{1-a-Interp-1-i-a}\nThis is a paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-2" '
            'data-label="1-a-Interp-1-i-a" id="1-a-Interp-1-i-a">'
            "This is a paragraph with a label.</p>",
        )

    def test_complicated_inline_interp_label(self):
        text = "{12-d-Interp-7-ii-c-A-7}\nThis is a paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-4" '
            'data-label="12-d-Interp-7-ii-c-A-7" id="12-d-Interp-7-ii-c-A-7">'
            "This is a paragraph with a label.</p>",
        )

    def test_prefixed_inline_interp_label(self):
        text = "{31-a-1-Interp-1}\nThis is a paragraph with a label."
        text2 = "{31-a-1-i-Interp-1}\nThis is a paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-0" '
            'data-label="31-a-1-Interp-1" id="31-a-1-Interp-1">'
            "This is a paragraph with a label.</p>",
        )
        self.assertEqual(
            regdown(text2),
            '<p class="regdown-block level-0" '
            'data-label="31-a-1-i-Interp-1" id="31-a-1-i-Interp-1">'
            "This is a paragraph with a label.</p>",
        )

    def test_empty_inline_interp_label(self):
        text = (
            "{30-Interp-1}\n"
            "1. Applicability\n\n"
            "{30-b-Interp}\n"
            "#### 30(b) Business Day"
        )
        self.assertIn(
            '<div class="regdown-block level-2" '
            'data-label="30-b-Interp" id="30-b-Interp"></div>',
            regdown(text),
        )

    def test_appendix_label(self):
        text = "{A-2-d}\nThis is a paragraph with a label."
        text2 = "{A-2-d-1}\nThis is another paragraph with a label."
        text3 = "{A-2-d-1-v}\nThis is yet another paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-0" '
            'data-label="A-2-d" id="A-2-d">'
            "This is a paragraph with a label.</p>",
        )
        self.assertEqual(
            regdown(text2),
            '<p class="regdown-block level-1" '
            'data-label="A-2-d-1" id="A-2-d-1">'
            "This is another paragraph with a label.</p>",
        )
        self.assertEqual(
            regdown(text3),
            '<p class="regdown-block level-2" '
            'data-label="A-2-d-1-v" id="A-2-d-1-v">'
            "This is yet another paragraph with a label.</p>",
        )

    def test_multi_part_appendix_label(self):
        text = "{M1-1-a}\nThis is a paragraph with a label."
        text2 = "{M1-1-a-1}\nThis is another paragraph with a label."
        text3 = "{M1-1-a-1-i}\nThis is yet another paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-0" '
            'data-label="M1-1-a" id="M1-1-a">'
            "This is a paragraph with a label.</p>",
        )
        self.assertEqual(
            regdown(text2),
            '<p class="regdown-block level-1" '
            'data-label="M1-1-a-1" id="M1-1-a-1">'
            "This is another paragraph with a label.</p>",
        )
        self.assertEqual(
            regdown(text3),
            '<p class="regdown-block level-2" '
            'data-label="M1-1-a-1-i" id="M1-1-a-1-i">'
            "This is yet another paragraph with a label.</p>",
        )

    def test_complex_appendix_label(self):
        text = "{B-h2-p2}\nThis is a paragraph with a label."
        text2 = "{B-h2-p2-1}\nThis is another paragraph with a label."
        text3 = "{B-h2-p2-1-i}\nThis is yet another paragraph with a label."
        self.assertEqual(
            regdown(text),
            '<p class="regdown-block level-0" '
            'data-label="B-h2-p2" id="B-h2-p2">'
            "This is a paragraph with a label.</p>",
        )
        self.assertEqual(
            regdown(text2),
            '<p class="regdown-block level-1" '
            'data-label="B-h2-p2-1" id="B-h2-p2-1">'
            "This is another paragraph with a label.</p>",
        )
        self.assertEqual(
            regdown(text3),
            '<p class="regdown-block level-2" '
            'data-label="B-h2-p2-1-i" id="B-h2-p2-1-i">'
            "This is yet another paragraph with a label.</p>",
        )

    def test_list_state(self):
        text = "- {my-label} This is a paragraph in a list."
        self.assertEqual(
            regdown(text),
            '<ul>\n<li>\n<p class="regdown-block level-1" '
            'data-label="my-label" id="my-label">'
            "This is a paragraph in a list."
            "</p>\n</li>\n</ul>",
        )

    def test_makeExtension(self):
        """Test that Markdown can load our extension from a string"""
        try:
            markdown.Markdown(extensions=["regdown"])
        except AttributeError as e:  # pragma: no cover
            self.fail(
                "Markdown failed to load regdown extension: "
                "{}".format(e.message)
            )

    def test_block_reference_resolver_not_callable(self):
        text = "see(foo-bar)"
        self.assertEqual(
            regdown(
                text,
                contents_resolver=None,
                render_block_reference=DEFAULT_RENDER_BLOCK_REFERENCE,
            ),
            "",
        )

    def test_block_reference_renderer_not_callable(self):
        text = "see(foo-bar)"
        self.assertEqual(regdown(text, render_block_reference=None), "")

    def test_block_reference_no_contents(self):
        contents_resolver = lambda l: ""
        text = "see(foo-bar)"
        self.assertEqual(
            regdown(
                text,
                contents_resolver=contents_resolver,
                render_block_reference=DEFAULT_RENDER_BLOCK_REFERENCE,
            ),
            "",
        )

    def test_block_reference(self):
        contents_resolver = lambda l: "{foo-bar}\n# §FooBar\n\n"
        text = "see(foo-bar)"
        self.assertIn(
            "<h1>§FooBar</h1>",
            regdown(
                text,
                contents_resolver=contents_resolver,
                render_block_reference=DEFAULT_RENDER_BLOCK_REFERENCE,
            ),
        )

    def test_tables_extension_exists(self):
        text = (
            "First Header  | Second Header\n"
            "------------- | -------------\n"
            "Content Cell  | Content Cell\n"
            "Content Cell  | Content Cell\n"
        )
        self.assertIn("<table>", regdown(text))
        self.assertIn("<th>First Header</th>", regdown(text))
        self.assertIn("<td>Content Cell</td>", regdown(text))

    def test_no_underscore_emphasis(self):
        self.assertIn("<em>foo</em>", regdown("*foo*"))
        self.assertIn("<strong>foo</strong>", regdown("**foo**"))
        self.assertIn("<strong><em>foo</em></strong>", regdown("***foo***"))
        self.assertNotIn("<em>foo</em>", regdown("_foo_"))
        self.assertNotIn("<strong>foo</strong>", regdown("__foo__"))
        self.assertNotIn("<strong><em>foo</em></strong>", regdown("___foo___"))

    def test_pseudo_form_field_end_of_line(self):
        text = "Form field: ___"
        self.assertIn(
            "Form field: "
            '<span class="regdown-form_extend">___'
            "<span></span></span>",
            regdown(text),
        )

    def test_pseudo_form_field_start_of_line(self):
        text = "__Form Field"
        self.assertIn(
            '<span class="regdown-form">__</span>Form Field', regdown(text)
        )

    def test_pseudo_form_field_inside_line(self):
        text = "inline______fields______within paragraph"
        self.assertIn(
            'inline<span class="regdown-form">______</span>'
            'fields<span class="regdown-form">______</span>'
            "within paragraph",
            regdown(text),
        )

    def test_pseudo_form_field_number_of_underscores(self):
        self.assertIn(
            '<span class="regdown-form_extend">__' "<span></span></span>",
            regdown("Field: __"),
        )
        self.assertIn(
            '<span class="regdown-form_extend">_______' "<span></span></span>",
            regdown("Field: _______"),
        )

    def test_section_symbol_with_non_breaking_space(self):
        self.assertIn("§&#160;1023", regdown("§ 1023"))
        self.assertIn("§&#160;1023.4", regdown("§ 1023.4"))
        self.assertIn("§&#160;1023.4(a)", regdown("§ 1023.4(a)"))
        self.assertIn("§&#160;1023.4(a)", regdown("§\t1023.4(a)"))
        self.assertIn("§&#160;1023.4(a)", regdown("§      1023.4(a)"))
        self.assertIn("§&#160;1023.4(a)", regdown("§&#160;1023.4(a)"))

    def test_emdash(self):
        self.assertIn("Some &mdash; text", regdown("Some --- text"))


class RegdownUtilsTestCase(unittest.TestCase):
    def test_extract_labeled_paragraph(self):
        text = (
            "{first-label} First para\n\n"
            "{my-label} Second para\n\n"
            "Third para\n\n"
            "{next-label}Fourth para"
        )
        result = extract_labeled_paragraph("my-label", text)
        self.assertIn("Second para", result)
        self.assertIn("Third para", result)

    def test_extract_labeled_paragraph_not_found(self):
        text = "{another-label} First para\n\n" "{next-label}Fourth para"
        result = extract_labeled_paragraph("my-label", text)
        self.assertEqual(result, "")

    def test_extract_labeled_paragraph_startswith(self):
        text = (
            "{first-label} First para\n\n"
            "{my-label} Second para\n\n"
            "Third para\n\n"
            "{my-label-1}Fourth para\n\n"
            "{my-label-2}Fifth para\n\n"
            "{next-label}Sixth para\n\n"
        )
        result = extract_labeled_paragraph("my-label", text, exact=False)
        self.assertIn("Second para", result)
        self.assertIn("Third para", result)
        self.assertIn("Fourth para", result)
        self.assertIn("Fifth para", result)


regdown/__init__.py
# -*- coding: utf-8 -*-
import re
from hashlib import sha3_224
from xml.etree import ElementTree as ET

from markdown import markdown, util
from markdown.blockprocessors import BlockProcessor, ParagraphProcessor
from markdown.extensions import Extension
from markdown.inlinepatterns import DoubleTagInlineProcessor, Pattern


# ***strongem*** or ***em*strong**
EM_STRONG_RE = r"(\*)\1{2}(.+?)\1(.*?)\1{2}"

# Form field: __
# __Form Field
# inline__fields__
PSEUDO_FORM_RE = r"(?P<underscores>_{2,50})(?P<line_ending>\s*$)?"

# Section symbol §
SECTION_SYMBOL_RE = r"(?P<section_symbol>§)\s+"

# Emdash ---
EMDASH_RE = r"---"


DEFAULT_URL_RESOLVER = lambda l: ""
DEFAULT_CONTENTS_RESOLVER = lambda l: ""
DEFAULT_RENDER_BLOCK_REFERENCE = (
    lambda c, **kwargs: "<blockquote>{}</blockquote>".format(regdown(c))
)


class RegulationsExtension(Extension):

    config = {
        "url_resolver": [
            DEFAULT_URL_RESOLVER,
            "Function to resolve the URL of a reference. "
            "Should return (title, url).",
        ],
        "contents_resolver": [
            DEFAULT_CONTENTS_RESOLVER,
            "Function to resolve the contents of a reference. "
            "Should return markdown contents of the reference or an empty "
            "string.",
        ],
        "render_block_reference": [
            DEFAULT_RENDER_BLOCK_REFERENCE,
            "Function that will render a block reference",
        ],
    }

    def extendMarkdown(self, md):
        md.registerExtension(self)

        # Replace all inlinePatterns that include an underscore with patterns
        # that do not include underscores.
        md.inlinePatterns.deregister("em_strong2")

        md.inlinePatterns.register(
            DoubleTagInlineProcessor(EM_STRONG_RE, "strong,em"),
            "em_strong2",
            60,
        )

        # Add inline emdash and pseudo form patterns.
        md.inlinePatterns.register(EmDashPattern(EMDASH_RE), "emdash", 200)
        md.inlinePatterns.register(
            PseudoFormPattern(PSEUDO_FORM_RE), "pseudo-form", 210
        )
        md.inlinePatterns.register(
            SectionSymbolPattern(SECTION_SYMBOL_RE), "section-symbol", 220
        )

        # Add block reference processor for `see(label)` syntax
        md.parser.blockprocessors.register(
            BlockReferenceProcessor(
                md.parser,
                url_resolver=self.getConfig("url_resolver"),
                contents_resolver=self.getConfig("contents_resolver"),
                render_block_reference=self.getConfig(
                    "render_block_reference"
                ),
            ),
            "blockreference",
            200,
        )

        # Replace the default paragraph processor with one that handles
        # `{label}` syntax and gives default hash-based ids to paragraphs
        md.parser.blockprocessors.deregister("paragraph")
        md.parser.blockprocessors.register(
            LabeledParagraphProcessor(md.parser), "paragraph", 10
        )

        # Delete the ordered list processor
        md.parser.blockprocessors.deregister("olist")


class EmDashPattern(Pattern):
    """Replace '---' with an &mdash;"""

    def handleMatch(self, m):
        return "{}mdash;".format(util.AMP_SUBSTITUTE)


class PseudoFormPattern(Pattern):
    """Return a <span class="pseudo-form"></span> element for matches of the
    given pseudo-form pattern."""

    def handleMatch(self, m):
        el = ET.Element("span")
        el.set("class", "regdown-form")
        if m.group("line_ending") is not None:
            el.set("class", "regdown-form_extend")
            ET.SubElement(el, "span")
        el.text = m.group("underscores")
        return el


class SectionSymbolPattern(Pattern):
    """Make whitespace after a section symbol non-breaking"""

    def handleMatch(self, m):
        return "{section}{stx}{char}{etx}#160;".format(
            section=m.group("section_symbol"),
            stx=util.STX,
            char=ord("&"),
            etx=util.ETX,
        )


class LabeledParagraphProcessor(ParagraphProcessor):
    """Process paragraph blocks, including those with labels.
    This processor entirely replaces the standard ParagraphProcessor in
    order to ensure that all paragraphs are labeled in some way."""

    RE = re.compile(r"(?:^){(?P<label>[\w\-]+)}(?:\s?)(?P<text>.*)(?:\n|$)")

    def test(self, parent, block):
        # return self.RE.search(block)
        return True

    def run(self, parent, blocks):
        block = blocks.pop(0)
        match = self.RE.search(block)

        if match:
            # If there's a match, then this is a labeled paragraph. We'll add
            # the paragraph tag, label id, and initial text, then process the
            # rest of the blocks normally.
            label, text = match.group("label"), match.group("text")
            # Labeled paragraphs without text should use a div element
            if text == "":
                el = ET.SubElement(parent, "div")
            else:
                el = ET.SubElement(parent, "p")
            el.set("id", label)
            el.set("data-label", label)

            # We use CSS classes to indent paragraph text. To get the correct
            # class, we count the number of dashes in the label to determine
            # how deeply nested the paragraph is. Inline interps have special
            # prefixes that are removed before counting the dashes.
            # e.g. 6-a-Interp-1 becomes 1 and gets a `level-0` class
            # e.g. 12-b-Interp-2-i becomes 2-i and gets a `level-1` class
            label = re.sub(
                r"^(\w+\-)+interp\-", "", label, flags=re.IGNORECASE
            )

            # Appendices also have special prefixes that need to be stripped.
            # e.g. A-1-a becomes a and gets a `level-0` class
            # e.g. A-2-d-1 becomes d-1 and gets a `level-1` class
            label = re.sub(r"^[A-Z]\d?\-\w+\-?", "", label)
            level = label.count("-")
            class_name = "regdown-block level-{}".format(level)
            el.set("class", class_name)

            el.text = text.lstrip()

        elif block.strip():
            if self.parser.state.isstate("list"):
                # Pass off to the ParagraphProcessor for lists
                super(ParagraphProcessor, self).run(
                    parent, blocks
                )  # pragma: no cover
            else:
                # Generate a label that is a hash of the block contents. This
                # way it won't change unless the rest of this block changes.
                text = block.lstrip()
                label = sha3_224(text.encode("utf-8")).hexdigest()
                class_name = "regdown-block"
                p = ET.SubElement(parent, "p")
                p.set("id", label)
                p.set("class", class_name)
                p.set("data-label", "")

                p.text = text


class BlockReferenceProcessor(BlockProcessor):
    """Process `see(label)` as an blockquoted reference.
    To render the block reference, the extension must be initialized with a
    callable render_block_reference option that will take the contents of
    the block reference, rendering to HTML, and return the HTML.
    """

    RE = re.compile(r"(?:^)see\((?P<label>[\w-]+)\)(?:\n|$)")

    def __init__(
        self,
        parser,
        url_resolver=None,
        contents_resolver=None,
        render_block_reference=None,
    ):
        super(BlockReferenceProcessor, self).__init__(parser)
        self.url_resolver = url_resolver
        self.contents_resolver = contents_resolver
        self.render_block_reference = render_block_reference

    def test(self, parent, block):
        return self.RE.search(block)

    def run(self, parent, blocks):
        block = blocks.pop(0)
        match = self.RE.match(block)

        if match:
            # Without a contents_resolver, we can't resolve block contents
            if not callable(self.contents_resolver) or not callable(
                self.render_block_reference
            ):
                return

            label = match.group("label")
            contents = self.contents_resolver(label)
            url = self.url_resolver(label)

            if contents == "":
                return

            rendered_contents = self.render_block_reference(contents, url=url)

            parent.append(ET.fromstring(rendered_contents.encode("utf-8")))


def makeExtension(*args, **kwargs):
    return RegulationsExtension(*args, **kwargs)


def regdown(text, **kwargs):
    """Regdown convenience function
    Takes text and parses it with the RegulationsExtention configured with
    the given keyword arguments."""
    return markdown(
        text,
        extensions=[
            "markdown.extensions.tables",
            "attr_list",
            RegulationsExtension(**kwargs),
        ],
        **kwargs,
    )


def extract_labeled_paragraph(label, text, exact=True):
    """Extract all the Regdown between the given label and the next.
    This utility function extracts all text between a labeled paragraph
    (with leading {label}) and the next labeled paragraph. If exact is False,
    it will match all paragraphs with labels that *begin with* the given
    label."""
    para_lines = []

    for line in text.splitlines(True):
        match = LabeledParagraphProcessor.RE.search(line)
        if match:
            # It's the correct label and we want an exact match
            if exact and match.group("label") == label:
                para_lines.append(line)

            # We don't want an exact match and the label starts with our given
            # label
            elif not exact and match.group("label").startswith(label):
                para_lines.append(line)

            # We've already found a label, and this is another one
            elif len(para_lines) > 0:
                break

        # We've found a label and haven't found the next
        elif len(para_lines) > 0:
            para_lines.append(line)

    return "".join(para_lines)
